### PR TITLE
Fix deprecated ReflectionParameter::getClass()

### PR DESCRIPTION
**Pull request by mvorisek**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | see title
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | n/a
| How to test?      | tests must pass
| Possible impacts? | no

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24669)

### CODE REVIEW

File: src/Core/Foundation/IoC/Container.php [hunk 1]
```diff
@@ -104,9 +104,8 @@ private function makeInstanceFromClassName($className, array $alreadySeen)
 
         if ($classConstructor) {
             foreach ($classConstructor->getParameters() as $param) {
-                $paramClass = $param->getClass();
-                if ($paramClass) {
-                    $args[] = $this->doMake($param->getClass()->getName(), $alreadySeen);
+                if ($param->getType() && !$param->getType()->isBuiltin()) {
```
Comment (matks) [inline src/Core/Foundation/IoC/Container.php]:
```
$param->getType()->isBuiltin()
```
this only works with ReflectionNamedType
https://www.php.net/manual/en/reflectionnamedtype.isbuiltin.php

Are we sure `$param->getType()` is a ReflectionNamedType? If yes I think PHPStan will require us to type-hint it

File: src/Core/Foundation/IoC/Container.php [hunk 2]
```diff
@@ -104,9 +104,8 @@ private function makeInstanceFromClassName($className, array $alreadySeen)
...
+                    $args[] = $this->doMake($param->getType()->getName(), $alreadySeen);
```
Comment (matks) [inline src/Core/Foundation/IoC/Container.php:109]: As reported by PHPStan, `ReflectionType::getName()` does not exist
https://www.php.net/manual/en/class.reflectiontype.php

It seems you should use `__toString()` instead 👍

**Review by matks — CHANGES_REQUESTED** [review]
Hello @mvorisek nice catch! Thanks for the PR.

PHPStan found a few mistakes can you fix them 😉

File: src/Core/Foundation/IoC/Container.php (hunk 2 above)
Comment (mvorisek) [reply src/Core/Foundation/IoC/Container.php:109]: Previously `$param->getType()` was checked if not null, now we check type 🚀

**Review by matthieu-rolland — DISMISSED** [review]

File: src/Core/Foundation/IoC/Container.php [hunk 3]
```diff
@@ -104,9 +104,8 @@ private function makeInstanceFromClassName($className, array $alreadySeen)
 
         if ($classConstructor) {
             foreach ($classConstructor->getParameters() as $param) {
-                $paramClass = $param->getClass();
-                if ($paramClass) {
-                    $args[] = $this->doMake($param->getClass()->getName(), $alreadySeen);
+                if ($param->getType() instanceof \ReflectionNamedType && !$param->getType()->isBuiltin()) {
```
Comment (Progi1984) [inline src/Core/Foundation/IoC/Container.php]: Could put the class `ReflectionNamedType` in `use` section ?

Comment (mvorisek) [reply src/Core/Foundation/IoC/Container.php]: Please do so and squash & merge.

Comment (Progi1984) [reply src/Core/Foundation/IoC/Container.php]: Please do it and I will take time to review your PR.

Comment (mvorisek) [reply src/Core/Foundation/IoC/Container.php]: Done, please take your time for my other PRs as well.

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
Code OK for me ✅ 
One more approval needed

Can probably by tested by a developer, wdyt @Robin-Fischer-PS ?

**Review by atomiix — APPROVED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hi @matks ! OK for QA by a dev :)

**Comment by mvorisek** [comment]
@matks closing for him as he close other legit PRs :D :D

**Comment by matks** [comment]
@mvorisek Actually I did not close this one because I think this one can move forward 😉  if however you wish to close it I understand

**Comment by kpodemski** [comment]
@mvorisek it's not like that, this PR brings additional value in eyes of PrestaShop maintainers, maybe other PRs not, we appreciate your work but that doesn't mean that we'll accept yours every single contribution to the project.

**Comment by mvorisek** [comment]
> 
> 
> @mvorisek it's not like that, this PR brings additional value in eyes of PrestaShop maintainers, maybe other PRs not, we appreciate your work but that doesn't mean that we'll accept yours every single contribution to the project.

@kpodemski indeed, it does :), can you do me a favor and review these:

- https://github.com/PrestaShop/PrestaShop/pull/22823
- https://github.com/PrestaShop/PrestaShop/pull/22965
- https://github.com/PrestaShop/PrestaShop/pull/23981

thanks!

**Comment by PierreRambaud** [comment]
> > @mvorisek it's not like that, this PR brings additional value in eyes of PrestaShop maintainers, maybe other PRs not, we appreciate your work but that doesn't mean that we'll accept yours every single contribution to the project.
> 
> @kpodemski indeed, it does :), can you do me a favor and review these:
> 
> * [Fix test DB prefixing, run SF2 update in the same php process #22823](https://github.com/PrestaShop/PrestaShop/pull/22823)
> * [Fix EOL to LF in multiline strings #22965](https://github.com/PrestaShop/PrestaShop/pull/22965)
> * [Fix the lowest composer dependencies #23981](https://github.com/PrestaShop/PrestaShop/pull/23981)
> 
> thanks!

I confirm these PRs bring no values compared to this one.
Please stop being rude when you speak to people, it will help you and people will maybe take time to review your PR next time ;)

**Comment by mvorisek** [comment]
@PierreRambaud you was not asked to talk as you do not understand them

**Comment by mvorisek** [comment]
let's ask @atomiix as well :)

**Comment by PierreRambaud** [comment]
> @PierreRambaud you was not asked to talk as you do not understand them

I clearly understand them, this is the reason why I don't want to merge them.
As I said above, your behavior is not correct, please don't forget our [Code of Conduct](https://www.prestashop-project.org/code-of-conduct/)

**Comment by mvorisek** [comment]
If two people can not agree, it is good to ask others without influencing them :)

**Comment by PierreRambaud** [comment]
> If two people can not agree, it is good to ask others without influencing them :)

We are talking a lot between maintainers, most of them agree but don't want to argue with you because of your behavior and how you interact with people.

**Comment by mvorisek** [comment]
Anyway, @PierreRambaud I see no review from you on https://github.com/PrestaShop/PrestaShop/pull/23981/files, what is wrong with this PR when you understand it?

**Comment by mvorisek** [comment]
> 
> 
> > If two people can not agree, it is good to ask others without influencing them :)
> 
> We are talking a lot between maintainers, most of them agree but don't want to argue with you because of your behavior and how you interact with people.

I am very kind, but the feedback needs to be constructive and moving us together to a fix/improvement/resolution.

**Comment by PierreRambaud** [comment]
FYI: https://github.com/PrestaShop/PrestaShop/pull/23969#issuecomment-870442975
I'm locking this Pull Request, even if you think you're very kind. Injure people from a certain country, ping people with only "Review and merge fast" is not constructive or help moving forward.

**Pull request closed without merging**